### ISSUE TITLE
Discussion: should we deprecate the old proxy event syntax?

### ISSUE THREAD

**Issue by Rich-Harris** [ractivejs/ractive#1172]
This follows on from the discussion on #1146 - see in particular feedback from @thomsbg (https://github.com/ractivejs/ractive/pull/1146#issuecomment-53526330) and @unity (https://github.com/ractivejs/ractive/pull/1146#issuecomment-54030188).

As of 0.5.6, we have two ways of dealing with events:

### Direct method call (new style)

```html
<button on-click='set("selected",foo)'>select foo</button>
```

### Triggering a proxy event, and handling it separately (old style)

```html
<button on-click='select:{{foo}}'>select foo</button>
```

```js
ractive.on( 'select', function ( event, selected ) {
  this.set( 'selected', selected );
});
```

These two are *functionally* identical, but *semantically* very different - the new syntax is far more concise and convenient, but introduces a coupling (between the event and the action) that wasn't there previously. For example, if you were using a Flux-style architecture, you wouldn't want to `set()` data directly.

## Option 1: deprecate

If we *did* deprecate the old syntax entirely, you *can* still maintain that decoupling:

```html
<!-- this... -->
<button on-click='fire("select",event,foo)'>select foo</button>

<!-- is identical to this: -->
<button on-click='select:{{foo}}'>select foo</button>
```

You can also call custom methods so that you're not blindly `set`ting data, but calling the same API that you'd call programmatically:

```js
var TodoList = Ractive.extend({
  add: function ( description ) {
    // any validation etc goes here...
    this.push( 'todos', { done: false, description: description });
  }
});
```

**Pros**:

* eliminates code from the codebase - makes it slimmer, more maintainable
* reduces the amount of stuff newcomers have to learn
* removes potential confusion
* we don't lose any expressive power

**Cons**:

* `on-click='fire("select",event,foo)'` is more verbose than `on-click='select:{{foo}}'`
* 'Proxy events' (i.e. 'click means select') are second-class citizens - some people prefer the proxy event style

## Option 2: Keep both syntaxes

**Pros**:

* Keeps existing users happy, means no breaking changes down the line
* Keeps proxy events as first-class citizens

**Cons**:

* Codebase size and maintainability
* Potentially confusing for newcomers

## Option 3: Keep old syntax, but disallow arguments

One idea we briefly tossed around was keeping `on-click='foo'` as a short version of `on-click='fire("foo",event)'`, but disallowing additional arguments. This alleviates the problem of situations like

```html
<widget on-foo='bar:{{these}},{{are}},{{ignored}}'>...</widget>
```

In that example, the `bar` handler receives the original `foo` event's arguments - the additional arguments are ignored. We don't have a particularly good way around this. Disallowing arguments masks this problem rather than solving it, but maybe there's some value in it.

**Pros/Cons**:

* Erm... not quite sure

## Option 4: Modify old syntax

If you think about it, the curly braces are fairly redundant. For data-binding, they serve as delimiters - it says...

```html
{{ <!-- references to the viewmodel are allowed in here --> }}
```

...but with method calls, the directive *itself* serves that purpose:

```html
<button on-click='doSomething( /* references are allowed in here */ )'>click me</button>
```

We could apply the same logic universally and just do away with the curlies...

```html
<button on-click='select:foo'>select foo</button>
```

...though for the sake of avoiding total confusion, it'd probably be better to go with something different, such as...

```html
<button on-click='select[foo]'>select foo</button>

<!-- If there we no arguments, syntax would be as currently is: -->
<button on-click='select'>select</button>
```

...which I personally find rather appealing. It gives us the opportunity to reverse an earlier mistake (whereby anything that couldn't be parsed was simply treated as a string - so `select:xyz` meant the same as `select:"xyz"` - bad API decisions are often made in the name of being 'forgiving' and 'flexible'. Mea culpa), and works equally well for decorators and transitions.

**Pros**:

* The 'array syntax' (let's call it that) is more compact than existing syntax
* Easy to parse using same mechanisms as method calls, so still gives us opportunities for code removal
* The arguments look like valid JavaScript! (Because they are)
* We don't need to do any data-binding - there's actually a small performance win with this approach (values only need to be got when the event actually happens), less state to manage and observe
* Proxy events are still first-class
* Adapts to decorators and transitions

**Cons**:

* Still involves two separate concepts for newcomers to learn
* Will be a breaking change when support for existing syntax is finally turned off

-----------------

So: would love to know what everyone thinks. Options 1 and 4 are my personal favourites, but your needs may well differ from mine. One thing I will say - I don't think we should worry about deprecating things. Ractive is still relatively young, and it's better that we take opportunities to get things right rather than trying to prevent the pain of updating existing templates. Deprecations can be well signposted with appropriate console warnings etc.

Thanks everyone.

**Comment by RangerMauve** [comment]
It'd be nice if it was more clear in the docs that `:` has a special meaning in event names. I was bashing my head at why `trigger:open` was only showing up as `trigger`.

**Comment by jarofghosts** [comment]
The new proxy event syntax is disappointingly similar to the native "on" attributes of yesteryear that I also find "repulsive" (re: the thread from a bit ago). I fail to see how this is a step forward and also how this "new" syntax can be considered "proxy events".

The "old" version was a very elegant and flexible solution to a common problem (and, in fact, was one of the things I liked so much about Ractive). For these reasons I humbly request that the syntax is not deprecated entirely, at least.

**Comment by matthew-dean** [comment]
@jarofghosts Agreed on both points. The ```fire()``` syntax is unnecessarily messy and obscures the function being called. The only thing I'd change is that the mustaching in the call seems unnecessary. That is:

```
<button on-click="select: {{ something }}">
```
should just be:
```
<button on-click="select: something">
```
There are no other "vars" / properties to reference here.

@Rich-Harris 
> the new syntax is far more concise and convenient

Nope.

**Comment by matthew-dean** [comment]
Oh, I see that that my suggestion was option 4. I really should read things.

**Comment by matthew-dean** [comment]
The other thing I could see doing is making an event reference optional. I think newbies would find this MUCH more comfortable:
```
<button on-click="select(something)">
```
Need the event passed? No worries, make a locally scoped variable ```e``` available for function calls.
```
<button on-click="select(e, something)">
```
Now it looks like everything a JavaScript developer is used to. And it should directly call functions within the ractive, rather than having to set ```ractive.on("select")``` The listener pattern doesn't really make sense (unless these events bubble through components, which may be the case - I'm not yet a Ractive expert).

**Comment by martypdx** [comment]
> unless these events bubble through components

Yep, that's the main reason to use events vs methods.

Given the introduction of `this.event`, I'd say there's even less of reason to need event arguments. 

So after almost a year, I have to say I now think Option 3 is the way to go. Use methods but allow the simple event synax: `on-click='foo'` as a short version of` on-click='fire("foo",event)'`

**Comment by matthew-dean** [comment]
> I'd say there's even less of reason to need event arguments.

Wait, what? I use event arguments all the time. How does this.event solve that?

**Comment by evs-chris** [comment]
I think he meant *the* `event` argument, not general event args.

**Comment by evs-chris** [comment]
I misread both the previous, apparently...

I'm for option 3 still too. Having `this.event` makes manually passing the special named event argument less painful because if you need it, it's just a context property away. It makes the transition from `foo:{{var1}},{{var2}}` to a method call a lot bit if you don't mind having event only methods on your instance (`foo(var1, var2)`). Literals are also a bit confusing in for args in the proxy syntax.

I also think fire should automatically propagate the event if it's called as a method event handler and there isn't an event argument in the appropriate position.

**Comment by thomsbg** [comment]
This issue is something I'd like to see a resolution to before the 0.8
release, one way or another.

After a year of working with ractive in a large application, I vote for any
option that has a unified and consistent syntax between event handlers,
decorators, and any other specially handled attribute. .

As much as possible, I'd like to limit the number of different syntaxes and
concepts in templates. Have an opinion and stick to it. If it means that
certain paradigms like proxying events become more verbose, than so be it.

At first, I thought the event proxying syntax was elegant and one thing
that made ractive different than other libraries (in a good way). But in
practice, I have made little use of event proxying.

90% of event handlers relate to behavior local to the component, so making
the 10% case more verbose is fine with me.

IMO, we should promote isolation between components over sharing data and
bubbled events. While useful tools, those concepts lead to messy code in
larger applications in my experience.

**Comment by Rich-Harris** [comment]
A few quick thoughts:

* Completely agree with @thomsbg that a unified syntax makes sense. The status quo (multiple ways of doing the same thing) is needlessly confusing
* From an implementation perspective – and I'm particularly sensitive to this after re-implementing it for #2022 – the current situation is *crazy*. There's a lot of code that just doesn't need to exist
* Again with my implementer's hat on, using curlies inside directives (`select:{{foo}},{{bar}}`) is infuriating. The reasons are historical – at the (very early) time proxy events were introduced, Ractive didn't have an expression parser, so using the data binding machinery made sense. But it's a hell of a lot easier if you bookend everything after the colon with `[`,`]` and run it through the expression parser – you get a function that (given the current values of `foo` and `bar`) simply returns the parameters
* Aesthetically, we're much better off without the curlies (e.g. `intro='fade:{delay:i*50}'` instead of the unreadable gubbins that happens when you mix object literals with mustache syntax)
* The method call syntax is the most flexible, and in my experience produces by far the most readable code (all the custom methods live in a single, obvious place, and a lot of the time you're calling built-in methods anyway! It's *much* more compact). When I need events to bubble, I find myself using `on-tap='fire("select",id)'` over `select:{{id}}` – it looks like regular JavaScript rather than some wacky DSL, and there's no weirdness about the `event` object being unshifted onto the list of arguments. (By the way, bubbling works fine in this scenario – http://jsfiddle.net/rich_harris/fvqew14a/)
* As a minor aside, components with custom methods work much better from a composition/inheritance perspective than components that depend on a lot of `this.on({...})` set-up work.

I know that not everyone feels the same way about the proxy event syntax, but I'd implore you to try and look at it from the perspective of someone new to Ractive. They can either use a familiar function call syntax, or learn a (let's face it, fairly weird) Ractive-specific way of doing things.

So in summary my inclination is basically option 1 – move over to the method-call syntax, and (gradually – *probably* not as part of 0.8) deprecate the old way.

### While I'm here

The one place where I'm continually frustrated with Ractive's event handling is with components:

```html
<!-- this is impossible... -->
<TodoList on-createTodo='persist(id,newTodo)'/>
```

We don't have a way to use the event arguments in the component's directive, which makes effective composition really difficult. It occurs to me that we could do something like this:

```html
<!-- supposing the `createTodo` event is fired with `id` and `newTodo` arguments -->
<TodoList on-createTodo='persist($1,$2)'/>

<!-- or if it's just fired with the `newTodo` argument -->
<TodoList on-createTodo='persist($1.id,$1)'/>
```

In other words, arguments like `$1` and `$2` are understood to have special values, the same way an `event` argument is replaced with `this.event` when the method is called.

**Comment by matthew-dean** [comment]
> I'd implore you to try and look at it from the perspective of someone new to Ractive

That's a good point; the colon / curly brace syntax of firing events took a bit to get used to. And direct method calls would be great for situations like:

```
<input type="checkbox" twoway="false" checked="{{ something < 2 }}" on-change="set(this.keypath, this.value ? 2: 0)" />
```
Or however that's written, but some way to deal with simple binding variations or type casting. (Incidentally, is that possible currently?)

**Comment by Rich-Harris** [comment]
@matthew-dean you can do this sort of thing:

```html
on-change="set('isChecked', event.node.checked ? 'yep' : 'nope')"
```

http://jsfiddle.net/rich_harris/4vqqca5h/

**Comment by matthew-dean** [comment]
@Rich-Harris Nice!!  ** Hurries to immediately implement **

**Comment by thomsbg** [comment]
Totally agree with all of those thoughts @Rich-Harris. 

I like the thought behind $1 and $2 arguments. It's not particularly pretty, but gets the job done. I wonder if would be useful to support $@ as the events arguments as an array. That combined with an ES6 style spread operator allows you to pass on all the arguments, without having to know how many there are.

    <todo-list on-newTodo="save(...$@)">

I usually turn my nose up at such perl-isms, maybe there's a more natural syntax.

**Comment by Rich-Harris** [comment]
@thomsbg I like the thinking – what about `on-newTodo="save(...arguments)"`? That's valid and idiomatic ES6.

**Comment by thomsbg** [comment]
Yeah totally. Seems obvious now :)

**Comment by Rich-Harris** [comment]
I should mention, for anyone pulling a O_o face at the `$1,$2` suggestion, that it *is* based on an existing JS convention :-)

```js
( 'this is a test string' ).replace( /this is a (\w+) (\w+)/, '$1 $2 $1 $2' );
// -> 'test string test string'
```

**Comment by martypdx** [comment]
I've written lots of boiler-plate code in last couple of months due to not having full component events args, so it's high on my priority list. I like the `...arguments`. Do we just stick with `arguments[0], arguments[1]` for individual access?

**Comment by LeJared** [comment]
I'd go with option 3. I use proxy events a lot (I like the separation of code (JS) and presentation (template)) and the eábillity to register multiple listeners on the same event. But I never ever use them with custom arguments.
It's just like you would use custom events in other libs (e.g. jQuery) most of the time and I find them really straightforward to use.

**Comment by matthew-dean** [comment]
@Rich-Harris Only after you commented on #2176 did I realize I was actually working with "new syntax" and "old syntax" simultaneously. No wonder I was confused. I thought all 3 function calls were slightly different with different use cases. I didn't really grok that one was replacing another.

I guess I use a mix of old and new, after learning I could do some of the "new". Passing arguments is easier with "new". Writing a bunch of methods to trigger is easier with "old" because I consider ```ractive.on``` a much cleaner and intuitive approach. That is, I would prefer to "define" using "old syntax" and "call" using new syntax. Either that or have an "on" property in the initial ractive instance definition (looks like that was mentioned by @martypdx).

One other thought, I think some of the visceral negative reaction to doing the smart thing and replacing "on-click" with "onclick" and calling methods directly, as it's "similar to yesteryear" is more about the reaction *then* then this suggestion now. The reaction against all that, and the cry for people to "separation content from presentation and content from behavior" was, on the one hand, a good idea for the maturity for the web platform, but in many, many cases, we threw the baby out with the bath water. There are plenty of times when specifying what update is going to take place or what's going to be fired as a result of the click is best in the template, just like sometimes it makes sense and is more maintainable to add inline styles. So, we're saying it's bad now, because it looks like then, and we said it was bad then, so it must be bad. But a lot of the reasoning for why it was "bad" was flawed at the time. It's actually *less* inutitive for newbies why Ractive insists that firing something on a click should be defined with "on-click" instead of "onclick". It's one of those places where people who know the internals of Ractive say, "Yeah, but they're different", to which I say that a newbie to Ractive cannot tell the difference.  So, for newbies, it's an arbitrary new syntax, even though it seems like it's doing the same thing: e.g. firing a method as a result of an interaction. And then later, when you learn that Ractive parses those elements and outputs something different anyway, the distinction between "on-click" and "onclick" actually feels more confusing and arbitrary.

So my suggestion: instead of avoiding what's familiar, build on what's familiar to web developers, and just make the behavior and result transparently better. Use what people know. Don't put up barriers to adoption because you have an aversion to an early stage of web development.

This is why syntaxes like Less and TypeScript (and the later syntax of Sass: SCSS) have flourished. They start with what people already know and use, and build on that knowledge in familiar ways.

Final thought: I think some of the reason there's repeated "inline" methods needed is that components are difficult to use, but that's probably a separate issue.

**Comment by LeJared** [comment]
Before deprecating anything, documentation and tutorials on ractive.org need to be updated.

The tutorials don't mention direct method call in events at all. At least they don't in the events section, where proxy events are the only type of event handling being taught.

Second, I've been search for more than an hour to find how so achive this: ``` on-click="toggle(@keypath+'.exclude')"```. I've not found a single example, which shows how to set a keypath in a direkt method call event handler, which is not a root key, not even in the discussion right here. All examples are too simplified to handle this, and at least to me the solution was not straight forward.

For proxy events this is show in many examples and as @matthew-dean mentioned above it's pretty familiar for developers coming from jQuery for example, at least without the ```event:{{prameter}}``` extra parameter syntax.

**Comment by TehShrike** [comment]
I came across this issue while looking to see if somebody else had reported https://github.com/ractivejs/ractive/issues/2303 already.

I've been using events lately as a way to set up observables and do FRP stream things.  I like events because I can write a template for a component, and then have generalized functions that I can use to compose things into streams.

For example, in my template I might do something like `on-change="user-did-thing-that-requires-server-interaction:thing,{{someObjectInScope}}"`.

In my code, say in my `complete` function, I can do something like `makeABaconJsStream(this, 'user-did-thing-that-requires-server-interaction').map(someOtherFunction).map(etc)`.

I can't replace this with function calls without changing my paradigm around quite a bit.

I value being able to pass a Ractive object to functions that can watch for certain events without them having to be concerned with anything but the events that will be emitted, and what the arguments will be.

**Comment by TehShrike** [comment]
My brain completely skipped over the `fire` usage somehow - after Rich enlightened me in https://github.com/ractivejs/ractive/issues/2303#issuecomment-162256308 I think I'm good to go with option 1.

I would suggest that using 
```
on-change="user-did-thing-that-requires-server-interaction:thing,{{someObjectInScope}}"
```

should produce the warning 

> event proxies are deprecated.  Replace with on-click="fire('user-did-thing-that-requires-server-interaction','thing',objectOnScope)"`

for maximum copy/paste handiness.

I am curious how long you would leave the warning in before removing the deprecated functionality?

**Comment by Rich-Harris** [comment]
Historically we've been pretty slow to remove deprecation warnings. If we got this into 0.8* it certainly wouldn't be removed before 0.9.

*getting it into 0.8 would consist of converting old-style event directives to new-style method calls at parse time. AFAICT the only cases where that would be impossible (or at least too difficult to bother with), i.e. a breaking change rather than a nuisance warning, is if you had sections in the directive, e.g. `on-click='foo:{{#each list}}{{this}}{{/each}}'`. Which I'd have thought was a pretty rare case

**Comment by evs-chris** [comment]
I've been looking at parsing for directives, mostly for decorators, but the comment in `processDirective` is not far off when it says 'shocking'. Between this issue, issues regarding the location of event handling functions, unification of syntax, and state support, I have yet another proposal to throw out.

### Cases

It seems that there are two flavors of directive parameters to accommodate: static naming and dynamic naming, both with optional arguments. For events, dynamic naming could be punted to the method flavo(u)r using `fire`, but that doesn't really work for decorators and transitions. So, I'll say that leaves both cases fully intact.

### Considerations

Method-style events, while much more convenient for arg-passing, have been the source of a fair amount of confusion regarding what exactly they can and will call. Some people expect them to bind like normal mustaches and call methods in the current context. They also look like a regular method call, so there is the temptation try to do things like `<a on-click="do.something(withParam).andThenCallMe()">`, which is not really unreasonable and _is_ entirely unsupported.

There's also parsing to consider. There is currently a bunch of funky special-cased logic in directive parsing that bends over backwards to parse stuff out of a string that has already been parsed as a plain text attribute value. I intend to address that with #2369, should that be accepted. While doing the refactor though, I'm going to have to keep a lot of _nuanced_ code around to handle legacy/deprecated cases.

Starting with 0.8, we will have a handy new special ref available that lets you get at Ractive instance methods as `@ractive.${method}`. @martypdx has opened #2345, which takes the special refs a step further and allows then `ractive.` portion to be left off, as long as what follows is not one of the reserved special refs.

Finally, I think Ractive is almost to the point of being feature complete for a 1.0 release. This is one of the last big things on _my_ list to get resolved before moving to get a 1.0 with longer-lived support out there. As it stands now, there has been so much churn between most releases that backporting fixes is nigh impossible without more bodies. I think that starting with 0.8 and its model refactor that there's not really another giant (practically) rewrite to be done. Also, the code has evolved into a structure that is easier to follow, which will hopefully be more attractive to contributors. Segue aside, I think it's time to knock out the breaking changes and get a final syntax out.

### Proposal

Given those cases and considerations, I would like to propose the we move entirely to two flavo(u)s of method-call syntax to handle events.

The first is a rework of static-naming that is currently handled by the existing method-call syntax. Instead of limiting this to a method on the Ractive instance, I propose that we make this a plain old expression and resolve and execute it when the triggering event fires. This would open event handling up to more styles programming with regards to data. For those like rich (fat) objects in their data, they could call methods on their models e.g. `{{#each users}}<a on-click=".disable()">Disable User</a>{{/each}}`. For those who like data-only models, they can call methods on their Ractive instance or even globals e.g. `<a on-click="@ractive.disableUser(.)">` (with #2345 `<a on-click="@disableUser(.)">`) or `<a on-click="@global.disableUser(.)">`. Since they're just expressions, this would also open up chained calls as mentioned above e.g. `<a on-click=".getHelper().doSomething(@keypath, 'some string')">`. Also, I just wanted to throw this out because moving an item in a list is currently more code than I think it should be: `<a on-click="@push('items', @splice('items', @index, 1).result[0])">`, where I've added a `result` property
to `Ractive.splice`'s returned Promise to avoid the asynchrony-for-transitions issue here.

Extra note for method-call events: they currently support passing args through via `$` params, and that would continue to be the case. In fact, I would add support for using `$` refs with the usual resolution process here (only for events) a la #2311.

The second, for dynamic names, is basically a variation of the existing proxy/decorator/transition syntax mangled slightly to make it easier to parse (both programmatically and in _my_ head). For events, this would always fire the event name with the `event` object supplied, as currently done by proxy events, followed by any args. The construct looks like this pseudo-BNF: `SIGIL(MUSTACHE|NAME)['('{ EXPRESSION, ',' }')']`. The `SIGIL` can be any character that makes sense. I propose `#` and perhaps even to make it configurable so that things that stand out more like 🔥(fire) could be used, perhaps as a list so that decorators and transitions could stand out too like 🎍(misc decoration) and 🏇(horse racing). The following `MUSTACHE|NAME` means that you can use a relaxed name for the event or you can resolve a string using a mustache. That is then followed by an optional args list in a regular function-callish form surrounded by parens. Here are the results of some of the various proxy scenarios (wrap each scenario in `<a on-click=${scenario}>` and feel free to swap `#` with 🔥, because unicode):

|Current|New|
|---|---|
|`"proxied"`|`#proxied"`|
|`"{{ foo ? foo.name : 'bar' }}"`|`"#{{ foo ? foo.name : 'bar' }}"`|
|`"proxied: stringArg"`|`#proxied('stringArg')`|
|`"proxied: {{ notStringArg }}"`|`#proxied(notStringArg)`|
|`"proxied: {{ 'stringArg' }}, {{ { some: { obj: 'settings' } } }}"`|"`#proxied('stringArg', { some: { obj: 'settings' } })`"|
|`"{{ foo ? foo.name : 'bar' }}: stringArg"`|`#{{ foo ? foo.name : 'bar' }}('stringArg')`|
|`"{{ foo ? foo.name : 'bar' }}: {{ notStringArg }}"`|`#{{ foo ? foo.name : 'bar' }}(notStringArg)`|
|`"{{ foo ? foo.name : 'bar' }}: {{ 'stringArg' }}, {{ { some: { obj: 'settings' } } }}"`|"`#{{ foo ? foo.name : 'bar' }}('stringArg', { some: { obj: 'settings' } })`"|

### Unification

Decorators and transitions also use the current proxy syntax for their names and args, and I would propose that they continue to do so with the new syntax. The one exception would be named decorators as implemented in #2369, where the name and parens would be left off, leaving just the args list e.g. `<div as-aceEditor>` and `<div as-aceEditor="'value', .options, 'otherArg'"></div><input type="checkbox" checked="{{.options.vimMode}}" /><input type="checkbox" checked="{{.options.wrap}}" />`. The current decorator syntax would also be supported e.g. `<div decorator="#aceEditor('value', .options, 'otherArg')">` for having dynamically named decorators.

Thanks for reading! I'd love to hear your thoughts on this because if it works, I'd like to go ahead an start a branch shooting for a 0.9 PR.

**Comment by Madgvox** [comment]
:+1:

**Comment by guilhermeaiolfi** [comment]
> I propose that we make this a plain old expression and resolve and execute it when the triggering event fires. 

YES!!! but the current ```on-click="method()"``` (without @) would be supported yet?

**Comment by evs-chris** [comment]
> but the current `on-click="method()"` (without @) would be supported yet?

Yes and no. Expressions resolve in the context in which they live, so without a special ref to link outside the data, there's no way to get to the ractive instance. I guess you could get funky and do something like:
```js
var r = new Ractive({...});
r.set('ractive', r);
```
I don't think that would destroy the world, but you would then have to prefix everything with wherever the instance was stored. Here're some examples for the proposed method-call replacement (static names and expressions):

```js
window.doSomething = function() { console.log('doing something global'); }

class User {
  constructor(name} { this.name = name; }
  sayHi() { console.log(`hi from ${this.name}`); }
}

var r = new Ractive({
  data: {
    doSomething() { return { action: { lastThing(arg) { console.log(`why? ${arg}`); } } }; },
    users: [
      new User('Joe'), new User('Rich'), new User('Marty')
    ]
  },
  doSomething() { console.log('doing something ractive-y'); }
});
```

```html
<button on-click="@global.doSomething()">log from the first function - global</button>
<button on-click="@ractive.doSomething()">log from ractive-y function</button>
with #2345 - <button on-click="@doSomething()">log from ractive-y function</button>
<button on-click="doSomething().action.lastThing('because I can')">log why? because I can from data</button>
{{#each users}}
  <button on-click=".sayHi()">will sayHi from current iteration user</button>
{{/each}}
<button on-click="@global.scrollTo(0, 1000)">scroll the window to 1000px</button>
```

Unfortunately, this doesn't make any fewer places to access stuff, but it does make it consistent with the rest of Ractive's mustachioed things. As mentioned above, it also frees up each person/team to pursue their preferred method of eventing.

**Comment by martypdx** [comment]
@evs-chris:

I really like the strict `@` prefixing for event methods. Yes it's breaking change, but then it's very consistent _and_ gives direct access to data functions. My one concern is that unless plumbed via adaptors, ractive won't know about stuff like `on-click='myList.add(thing)'` ). With more proxy and less ie8 this seems the way to head, but could end up catching people off guard.

Also, I'm in agreement that we should broaden the use of in-template scripting with more complex expressions.

I would rather we not introduce an additional invocation api for decorators and transitions. If I understand correctly, the issue is to move the token with the attribute value can be dynamically secified?.

I would suggest a few (non-mutually exclusive) ideas:

1) Use the conditional attribute blocking:

```html
<div {{#if shouldDecorate}}as-sortable='{ someOption: true}'{{/if}} 
     {{#if foo}} as-tooltip {{else}} as-ariaplus {{/if}} ></div>
```
2) Use `[transition]-in`, `[transition]-out`, and `[transition]-in-out` as replacements for `intro`, `outro` and `intro-outro` and same block semantics apply:

```html
<div {{#if foo}}fade-in-out{{/if}}></div>
```

3) Only allow the shortcut of single name token to fire an event implicitly:
```html
<div on-click='submit'></div>
<!-- shortcut for -->
<div on-click='@fire( "submit", event )'></div>
```
I like the consistency over what seems like a rather specialized use of `#` for doing `#handler( someData )` in lieu of `@fire( "handler", someData )` (or more specifically `@fire( "handler", event, someData )` if you didn't want to use `this.event`).

However an additional idea I had for the `#` would be to use it as a general sigil for a "dynamic directive". Perhaps using `#(expression)` for expressions:

```html
<!-- uses value of foo, view, eventName and transitionName: -->
{{> #foo }}
<#view user='{{user}}'/>
<div on-#eventName='@process()' #transitionName-in-out></div>

<!-- uses result of expression for partial, component, and event names: -->
{{> #(field.type + 'field') }}
<#( mobile ? 'sidebar' : 'navbar') user='{{user}}'/>
<div on-#(foo ? 'click' : 'mousedown')='@process()'></div>
```

**Comment by evs-chris** [comment]
I'm for 1 and 2, too. With directives allowed in conditionals, it seems like the need for dynamic names would diminish significantly. It'd be good to hear a few opinions from those who have actually used dynamically named decorators or transitions though, as I have not.

Dropping dynamic names would allow us to use the expression parser directly inside of directives e.g. `<div as-foo="., 'some string', [{ another: 'arg'}]" fade-in="{ delay: 500 }"></div>`, which is basically what you said in 1. But that is the sticky issue with parsing for directives right now: one attribute value is being used to handle three cases. Stripping out the dynamic names drops it to two more easily handled cases - plain attributes and args lists for directives.

For 3, I'm starting to think that maybe a library/plugin solution is in order. Since `this.event` is available, perhaps a `proxy` method could be added that just refires the event keeping the params intact. It's slightly longer, but there's no longer a special case:
```html
<div on-click="@proxy('submit')"></div>
<!-- instead of the slightly shorter, but much more special -->
<div on-click="submit"></div>
```

For part 2 of number 3, that's pretty slick. I'd like to throw out that perhaps inline template strings might be a good solution. I'd eventually like to get them supported in expressions, but there's no reason they couldn't also be used in templates for dynamic bindings in the future:

```html
<div class="{{`here is an ${expression.interpolated.template} because yay es2015!`}}"></div>
<div `dynami${foo ? 'c' : 'ism'}Ftw`="{{`this is ${foo ? 20 : 60`}}"></div>
```

So basically, if the template string is encountered outside of a mustachioed region, it means "this is a dynamic binding". That would come back around to solving numbers 1 and 2, though I would punt on that for now.

**Comment by arxpoetica** [comment]
Some parsers (Jade, Slim) will not play nicely with `#` in inline code such as `<div on-#eventName='@process()' #transitionName-in-out></div>`, so can we think of something else? Plus, it looks kind of gross, hanging on the attribute name like that.

**Comment by guilhermeaiolfi** [comment]
Let's keep in mind that every symbol/behavior added will make it harder to beginners to learn ractive and more things to remember for us. A small and efficient API is more important in the long run. Based on that, I would say to deprecated the old style (proxy) and make it a plain old expression. That would open the door to create something like @evs-chris suggested (``` @proxy('submit')```) or other (crazier) ways to do it. I went to see how other frameworks do it:

1) Vue: it doesn't even support something different than method calls in templates: [Custom Events](http://vuejs.org/guide/components.html#Custom_Events)

2) react.js: the whole thing is a mess. Hard to understand the popularity of it.

3) ember: it seems that the only way of doing it is triggering it inside a method: http://emberjs.com/api/classes/Ember.Evented.html

**Comment by RangerMauve** [comment]
Personally, I'm phasing out the use of events provided by ractive and implementing a more simple system over top because there's a lot of syntax to deal with. One problem was that I had events with colons in their names already in use before we switched to ractive, so I had to add a conversion step between ractive events and everything else. Working with events coming from components has also been a bit of a pain.

**Comment by JonDum** [comment]
I'm sort of scared of all these new `@`, `#` symbols being proposed. Yes, they offer crazy amounts of flexibility but at the huge cost of complexity—both in the mental model of users and the code base. 

When I was first learning about ractive I loved that it _didn't_ have a bunch of symbols that I would have needed to learn exactly what each one did. All these symbols make ractive's templates look quite a bit more daunting and scary to the newcomer.

I think @Rich-Harris's original Option 1 (method calls) is perfectly fine for 90% of use cases and should be promoted as the go-to way of doing.

That leaves us with two clean and easy to understand syntaxes

```html
<div on-click="toggle('foo')">

<div on-click="doSomething(foo)"/> 
```
where `foo` in `doSomething(foo)` gets resolved as an expression (`{{foo}}`)

The other syntaxes are sometimes convenient, but more often than not are very clunky and hard to move around, like Rich mentioned.

---

Now for this `@` stuff... couldn't you achieve the same result with the existing non-ambiguous reference syntax? Then we wouldn't be introducing new symbols, but using the same mental model from previously introduced ones.

` `: (no prefix) ambiguous resolve
`.`: current scope/iteraton (data)
`~/`: current prototype
`/`: global

Then you'd have the same control over which method is called as you do in expressions.

```js
window.doSomething = function() { console.log('doing something global'); }

class User {
  constructor(name} { this.name = name; }
  sayHi() { console.log(`hi from ${this.name}`); }
}

var r = new Ractive({
  data: {
    doSomething() { return { action: { lastThing(arg) { console.log(`why? ${arg}`); } } }; },
    users: [
      new User('Joe'), new User('Rich'), new User('Marty')
    ]
  },
  doSomething() { console.log('doing something ractive-y'); }
});
```
```html

<!-- call the doSomething from the prototype -->
<button on-click="~/doSomething()">log from ractive-y function</button>

<!-- call the doSomething in data: {} -->
<button on-click=".doSomething().action.lastThing('because I can')">log why? because I can from data</button>
or
<button on-click="doSomething().action.lastThing('because I can')">also works because same method on data and prototype, prefers data</button>

<!-- call the sayHi from the model -->
{{#each users}}
  <button on-click=".sayHi()">will sayHi from current iteration user</button>
{{/each}}

<!-- do the global -->
<button on-click="/scrollTo(0, 1000)">scroll the window to 1000px</button>
```

**Comment by evs-chris** [comment]
@JonDum as of my last response to @martypdx, I'm thinking the `#`s should go away in the context of this issue.

The special refs with `@` are meant to be an evolution of the same concept from Handlebars, where `@index` is the index of the current iteration, etc. We recently added `@ractive`, which is a reference to the current Ractive instance, and `@glboal`, which is a reference to the `window` or `global` object depending on the environment. @martypdx has a PR open that basically shortens `@ractive.some.path` to just `@some.path` to drop a little of the noise. This gives easy access to non-data state (props, if you will) for those who don't want to taint their data with, for instance, helper methods or third party library handles.

Unfortunately, you're changing the meaning of `~/` in the context of events to mean `@ractive`, whereas currently it means "the root of the data for this instance" anywhere else. You're also adding a new thing, `/`, to mean `@global`.

There has also been a fair amount of misunderstanding of the difference between these and why you can't call the `doSomething` method from the latter in an event:
```js
new Ractive({
  data: { ... },
  doSomething() { ... }
})

new Ractive({
  data: {
    doSomething() { ... }
  }
})
```

Ignoring the proxy issue (`#`s) from my initial proposal, it is an attempt to unify how the various resolutions from data references and events work by simply making them expressions that are executed when the event fires. I was hoping that was less surprising than having to switch contexts just for calling event handlers. To me, at least, it is much more obvious what is going to happen with `<button on-click="@ractive.set('foo', 42)">` vs `<button on-click="set('foo', 42)">`.

In the docs overhaul (I'm bout to start, I swear), having a good once-over on what keypaths can look like and having them be the same no matter where they occur would hopefully reduce beginner intimidation. Then we're just down to normal data refs and special refs.

**Comment by akhoury** [comment]
@evs-chris 

>  you can't call the doSomething method from the latter in an event:

yea you can :P lol 

```javascript
this.on(*, function(re) {
    if (re && re.original && re.original.preventDefault && re.node
        && ((window.HTMLElement && re.node instanceof window.HTMLElement) || re.node instanceof window.Element) 
        && re.name && typeof this[re.name] === 'function') {
          
        return this[re.name].apply(this, arguments);
    }
} 
```

**Comment by fskreuz** [comment]
> I went to see how other frameworks do it: @guilhermeaiolfi 
> I'm sort of scared of all these new @, # symbols being proposed. @JonDum 

Don't forget Angular and its `=`, `&` and `@` on directives. So terrible that every time I revisit an Angular project, I have to read up which does what. 

Been a long time since I was on this thread, and since then, I have settled with the method call syntax. It's reminiscent of the good ol' inline vanilla inline script calls. A setup like the following is a no-brainer for a newcomer.

A call from an event (the `on-*`) is a method call

```xml
<button on-click="someMethod(arg1, arg2, ...)">Hi!</button>
<component on-click="someMethod(arg1, arg2, ...)">Hi!</component>
```

A call from a `decorator` is a decorator setup
```xml
<span decorator="i18n(arg1, arg2, ...); lowercase(); ">Hi!</span>
```

A call from a `transition` is a transition
```xml
<div transition="fade(arg1, arg2, ...); slide(arg1, arg2)">Hi!</div>
```

Anything between `{{ }}` prints whatever is evaluated into the template. They resolve to data/computed props. Expressions allowed.
```xml
<!-- primitive -->
<p>{{ model.summary }}</p>

<!-- data model (covers cases where data contains functions like getters, setters) -->
<p>{{ model.getSummary() }}</p>

<!-- expressions -->
<p>{{ foo ? 'bar' : 'baz' }}</p>
```

For references/shorthands, I suggest verbose naming instead of standalone `#`, `@`, `.` etc. 

- `@global` - The global object (`window` on the browser)
- `@instance` - The current instance. All the stuff defined there accessible via this.
- `@root` - The top-level component, same characteristics as `@instance`.
- `@parent` - The parent of the current instance, same characteristics as `@instance`.
- `@whatever` - as long as it's not a solo doodad.

**Comment by evs-chris** [comment]
@akhoury I can't really tell what is happening there, but I'll go with 'I stand corrected' :laughing: 

@fskreuz are those changes you'd like to see, or were you just examining it from a beginner's viewpoint? Just wondering on a few points, cause I'm gathering ideas for a PR:

1. I'm not opposed to `;` delimited expressions, but I've seen opinions expressed in issues that were strongly opposed. I think it'd be kinda nice for event handlers where you want to do two or three quick and simple things like a splice, a toggle, and a set.
2. Would you stick strictly to calling only Ractive instance methods from event handlers?
3. How do you feel about `as-${decorator}`, like `as-lowercase` and `as-i18n="arg1, arg2, ..."`, and the proposed `${transition}|-in`, like `slide-in`, `fly-out`, `fade-in-out`, and `fade-in="{ duration: 500 }"`?

We could also start using `%=` on attributes to indicate transclusion during the next digest cycle after the isomorphic cloud synergizes. :wink:

**Comment by fskreuz** [comment]
> are those changes you'd like to see, or were you just examining it from a beginner's viewpoint? 

Both? :smile: I fear that all these Ractive-specific syntax makes it difficult to board devs, given that Ractive has remained faithful to its 99% vanilla look with a touch of Handlebars.

For comparison, ES6+JSX. I have this belief that most of the people who like the React+JSX combo was because it was eerily similar to PHP+[Heredoc](https://secure.php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc).

```php
<?php
// This looks like a React component now does it? But it's PHP!
class MyComponent extends Component{
  public function render(){
    $message = 'world';
    return <<<HEREDOC
      <div>Hello {$message}!</div>
HEREDOC;
  }
}
```

So if Ractive stuck to the vanilla HTML+JS syntax in templates, it's a win. One doesn't need to unlearn anything, change their mental model and so forth. 

> I think it'd be kinda nice for event handlers where you want to do two or three quick and simple things like a splice, a toggle, and a set.

> How do you feel about as-${decorator}, like as-lowercase and as-i18n="arg1, arg2, ...", and the proposed ${transition}|-in, like slide-in, fly-out, fade-in-out, and fade-in="{ duration: 500 }"?

I think there was a discussion on multiple decorators and stuff like that. Was wondering if we could have a common syntax for all. All method-like, all semi-colon-delimited, all supporting multiple operations and sequential operations.

Multiple parallel calls could use semi-colon delimited like multiple inline JS calls.

```xml
<div transition="slide({ ... }); fade({ ... })"></div>
```

For sequential, dot-delimited. It's just like what you'd expect from libraries like jQuery and Lodash.

```xml
<!-- sequential animation (notice the dot between, as if it was chained) -->
<div transition="slide({ ... }).fade({ ... })"></div>
```

`as-*` wouldn't really tell me that it's a decorator. Probably `decorator-*`. But then, this means they will either look like `prefix-name="arg1, arg2,..."` (which is similar to the old syntax) or `prefix-name="{ arg1: val1, ... }"` which becomes more Angular-y. :scream: And both of which are departures from the method-like syntax of method calls.

> Would you stick strictly to calling only Ractive instance methods from event handlers?

This I'm still thinking about. It does make sense in the following:

```xml
<button on-click="spamAds(100, emails)">{{ @instance.translate('Hello') }}</button>
<button on-click="spamAds(100, emails)">{{ @methods.translate('Hello') }}</button>
```

`@instance` and `@methods` tell me we're referencing the instance and not the data, making `translate` not a data function. As long is it looks as verbose as this, and the docs telling people that by default `{{ }}` references data, and you need to use `@` to reference stuff other than data, then it's cool.

> We could also start using %= on attributes to indicate transclusion during the next digest cycle after the isomorphic cloud synergizes. 

![image](https://cloud.githubusercontent.com/assets/1582932/12918568/cca55060-cf0d-11e5-8710-16e8b2f16817.png)

http://dilbert.com/strip/2015-12-20

**Comment by dagnelies** [comment]
TL;DR

this is getting out of hands. Sorry. Too much talk. Just *one* begging: please keep it simple! ...there is already too much around!

**...therefore, I'm still in favor of option 1. Please just deprecate the old proxy syntax, it ain't needed anymore.**

As for all other proposals, I'm in favor of putting them in separate threads, it's getting too messy/overcrowded here IMHO.

**Comment by heavyk** [comment]
if you really want to growth hack the analytics, perhaps a better way, would be to make a babel transform which reads the developer's mind.

**Comment by codler** [comment]
I like option 1 and 4, deprecate old syntax and also try the "array syntax"

**Comment by tomekmarchi** [comment]
Remove the old syntax. One con that is, no pun intended, a growing concern is file size. When more important features find their way in removing useless dead code is mandatory IMO. Force people to change simple as that of they don't like it that's on them. I rather see a decision in favor of the ecosystem not a select few.

**Comment by kouts** [comment]
Starting with Ractive nearly 4 months ago, I thought proxy events is the "right way" to call a method.
In http://docs.ractivejs.org/latest/method-calls clearly states that "As an **alternative** to proxy events, you can call any method...". So in my mind  the "method call" stayed as the alternative syntax not the main syntax. As a result my code is now full of 
```
on-click="add_item" 
on-click="delete_item"
```
etc...

I use method calls in templates only when I want to use some ractive methods directly e.g set, push, splice and so on...

If you are planning of removing the proxy syntax you may want to change the docs so that it is clear for newcomers that the "Method calls" syntax is the proposed one and not the other way around.

With that said I agree with @maxgalbu and say that: I'd go with option 1, but without the mandatory event parameter (if possible).

**Comment by TehShrike** [comment]
Yeah, since I've learned that proxy events are deprecated for most intents and purposes, I feel like I need to warn everyone when I tell them about Ractive.  "It's great and you should use it but remember to use this `fire('event')` syntax for events instead of what the docs say"

**Comment by arisAlexis** [comment]
I use mostly method calls but sometimes proxy events are the way to go because you may not want to call a method on the instance but propagate an event on the parent view. Otherwise you need to call a method which will call this.fire which is extra code and kind of unecessary. Am I missing something?

**Comment by guilhermeaiolfi** [comment]
> Otherwise you need to call a method which will call this.fire which is extra code and kind of unecessary. Am I missing something?

There is no need to call a method that will fire your event. You would call ```@this.fire('eventName', event, data)``` in your template and that's it.
